backend/de.metas.business/src/main/java/de/metas/invoice/interceptor/C_Invoice.java

```diff
@@ -86,24 +86,26 @@ public C_Invoice(@NonNull final PaymentReservationService paymentReservationServ
 		this.paymentReservationService = paymentReservationService;
 	}
 
-	@DocValidate(timings = { ModelValidator.TIMING_BEFORE_COMPLETE })
-	public void onBeforeComplete(final I_C_Invoice invoice)
-	{
-		allocateInvoiceAgainstCreditMemo(invoice);
-		linkInvoiceToPaymentIfNeeded(invoice);
-	}
-
 	@DocValidate(timings = { ModelValidator.TIMING_AFTER_COMPLETE })
 	public void onAfterComplete(final I_C_Invoice invoice)
 	{
-		markAsPaid(invoice);
+		testAndMarkAsPaid(invoice);
+		allocateInvoiceAgainstCreditMemo(invoice);
+		linkInvoiceToPaymentIfNeeded(invoice);
 		allocateInvoiceAgainstPaymentIfNeeded(invoice);
+		autoAllocateAvailablePayments(invoice);
 		captureMoneyIfNeeded(invoice);
 		ensureUOMsAreNotNull(invoice);
 
 		C_Invoice_CreateExportData.scheduleOnTrxCommit(invoice);
 	}
 
+	private void autoAllocateAvailablePayments(final I_C_Invoice invoice)
+	{
+		allocationBL.autoAllocateAvailablePayments(invoice);
+		testAndMarkAsPaid(invoice);
```

Comment:
why do we have to call testAndMarkAsPaid again. we are already calling it at line 92....

I think we shall try calling it only once.